{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **Novozymes-Enzyme-Stability-Prediction**\n",
    "\n",
    "##### Develop models that can predict the ranking of protein stability (as measured by melting point, ```tm```) after single-point amino acid mutation and deletion. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Import Libraries**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from io import StringIO\n",
    "import biopandas\n",
    "from biopandas.pdb import PandasPdb\n",
    "pdb = PandasPdb()\n",
    "\n",
    "import tensorflow as tf\n",
    "import pandas as pd\n",
    "pd.options.mode.chained_assignment = None\n",
    "pd.set_option('display.max_columns', None)\n",
    "\n",
    "import numpy as np\n",
    "import sklearn\n",
    "from collections import Counter\n",
    "from datetime import datetime\n",
    "from zipfile import ZipFile\n",
    "from glob import glob\n",
    "import Levenshtein, warnings, requests, hashlib\n",
    "import imageio, IPython, sklearn\n",
    "import urllib, zipfile, pickle, random\n",
    "import shutil, string, json, math, time, gzip\n",
    "import ast, sys,io, os, gc, re\n",
    "\n",
    "from matplotlib.colors import ListedColormap\n",
    "from matplotlib.patches import Rectangle\n",
    "import matplotlib.patches as patches\n",
    "import plotly.graph_objects as go\n",
    "import matplotlib.pyplot as plt\n",
    "from tqdm.notebook import tqdm; tqdm.pandas();\n",
    "import plotly.express as px\n",
    "import tifffile as tif\n",
    "import seaborn as sns\n",
    "from PIL import Image, ImageEnhance; Image.MAX_IMAGE_PIXELS = 5_000_000_000;\n",
    "import matplotlib\n",
    "from matplotlib import animation, rc\n",
    "rc('animation', html='jshtml')\n",
    "\n",
    "import plotly, PIL, cv2\n",
    "import plotly.io as pio\n",
    "\n",
    "def seed_all(seed=7):\n",
    "    os.environ['PYTHONHASHSEED'] = str(seed)\n",
    "    random.seed(seed)\n",
    "    np.random.seed(seed)\n",
    "    tf.random.set_seed(seed)\n",
    "\n",
    "seed_all()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Helper Functions**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_mutation_info(_row, _wildtype=\"VPVNPEPDATSVENVALKTGSGDSQSDPIKADLEVKGQSALPFDVDCWAILCKGAPNVLQ\" \\\n",
    "                                      \"RVNEKTKNSNRDRSGANKGPFKDPQKWGIKALPPKNPSWSAQDFKSPEEYAFASSLQGGT\" \\\n",
    "                                      \"NAILAPVNLASQNSQGGVLNGFYSANKVAQFDPSKPQQTKGTWFQITKFTGAAGPYCKAL\" \\\n",
    "                                      \"GSNDKSVCDKNKNIAGDWGFDPAKWAYQYDEKNNKFNYVGK\"):\n",
    "    terminology_map = {\"replace\":\"substitution\", \"insert\":\"insertion\", \"delete\":\"deletion\"}\n",
    "    req_edits = Levenshtein.editops(_wildtype, _row[\"protein_sequence\"])\n",
    "    _row[\"n_edits\"] = len(req_edits)\n",
    "    \n",
    "    if _row[\"n_edits\"]==0:\n",
    "        _row[\"edit_type\"] = _row[\"edit_idx\"] = _row[\"wildtype_aa\"] = _row[\"mutant_aa\"] = pd.NA\n",
    "    else:\n",
    "        _row[\"edit_type\"] = terminology_map[req_edits[0][0]]\n",
    "        _row[\"edit_idx\"] = req_edits[0][1]\n",
    "        _row[\"wildtype_aa\"] = _wildtype[_row[\"edit_idx\"]]\n",
    "        _row[\"mutant_aa\"] = _row[\"protein_sequence\"][req_edits[0][2]] if _row[\"edit_type\"]!=\"deletion\" else pd.NA\n",
    "    return _row\n",
    "\n",
    "def revert_to_wildtype(protein_sequence, edit_type, edit_idx, wildtype_aa, mutant_aa):\n",
    "    if pd.isna(edit_type):\n",
    "        return protein_sequence\n",
    "    elif edit_type!=\"insertion\":\n",
    "        new_wildtype_base = protein_sequence[:edit_idx]\n",
    "        if edit_type==\"deletion\":\n",
    "            new_wildtype=new_wildtype_base+wildtype_aa+protein_sequence[edit_idx:]\n",
    "        else:\n",
    "            new_wildtype=new_wildtype_base+wildtype_aa+protein_sequence[edit_idx+1:]\n",
    "    else:\n",
    "        new_wildtype=protein_sequence[:edit_idx]+wildtype_aa+protein_sequence[edit_idx:]\n",
    "    return new_wildtype\n",
    "\n",
    "def flatten_l_o_l(nested_list):\n",
    "    \"\"\" Flatten a list of lists \"\"\"\n",
    "    return [item for sublist in nested_list for item in sublist]\n",
    "\n",
    "def print_ln(symbol=\"-\", line_len=110):\n",
    "    print(symbol*line_len)\n",
    "    \n",
    "# Note mutation edit_idx is offset by 1 as many tools require it to be 1 indexed to 0 indexed.\n",
    "def create_mutation_txt_file(_test_df, filename=\"/kaggle/working/AF70_mutations.txt\", return_mutation_list=False, include_deletions=False):\n",
    "    if return_mutation_list: mutation_list = []        \n",
    "    with open(filename, 'w') as f:\n",
    "        for _, _row in _test_df[[\"protein_sequence\", \"edit_type\", \"edit_idx\", \"wildtype_aa\", \"mutant_aa\"]].iterrows():\n",
    "            if not include_deletions and (pd.isna(_row[\"edit_type\"]) or _row[\"edit_type\"]==\"deletion\"): continue\n",
    "            f.write(f'{_row[\"wildtype_aa\"]+str(_row[\"edit_idx\"]+1)+(_row[\"mutant_aa\"] if not pd.isna(_row[\"mutant_aa\"]) else \"\")}\\n')\n",
    "            if return_mutation_list: mutation_list.append(f'{_row[\"wildtype_aa\"]+str(_row[\"edit_idx\"]+1)+(_row[\"mutant_aa\"] if not pd.isna(_row[\"mutant_aa\"]) else \"\")}')\n",
    "    if return_mutation_list: return mutation_list\n",
    "        \n",
    "def create_wildtype_fasta_file(wildtype_sequence, filename=\"/kaggle/working/wildtype_af70.fasta\"):\n",
    "    with open(filename, 'w') as f: f.write(f\">af70_wildtype\\n{wildtype_sequence}\")\n",
    "        \n",
    "def uniprot_id2seq(uniprot_id, _sleep=3):\n",
    "    base_url = \"http://www.uniprot.org/uniprot\"\n",
    "    full_url = os.path.join(base_url, str(uniprot_id)+\".fasta\")\n",
    "    _r = requests.post(full_url)\n",
    "    if _r.status_code!=200: print(_r.status_code)\n",
    "    if _sleep!=-1: time.sleep(_sleep)\n",
    "    return ''.join(_r.text.split(\"\\n\")[1:])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Loading Data**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>protein_sequence</th>\n",
       "      <th>pH</th>\n",
       "      <th>data_source</th>\n",
       "      <th>tm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>AAAAKAAALALLGEAPEVVDIWLPAGWRQPFRVFRLERKGDGVLVG...</td>\n",
       "      <td>7.0</td>\n",
       "      <td>doi.org/10.1038/s41592-020-0801-4</td>\n",
       "      <td>75.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>AAADGEPLHNEEERAGAGQVGRSLPQESEEQRTGSRPRRRRDLGSR...</td>\n",
       "      <td>7.0</td>\n",
       "      <td>doi.org/10.1038/s41592-020-0801-4</td>\n",
       "      <td>50.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>AAAFSTPRATSYRILSSAGSGSTRADAPQVRRLHTTRDLLAKDYYA...</td>\n",
       "      <td>7.0</td>\n",
       "      <td>doi.org/10.1038/s41592-020-0801-4</td>\n",
       "      <td>40.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>AAASGLRTAIPAQPLRHLLQPAPRPCLRPFGLLSVRAGSARRSGLL...</td>\n",
       "      <td>7.0</td>\n",
       "      <td>doi.org/10.1038/s41592-020-0801-4</td>\n",
       "      <td>47.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>AAATKSGPRRQSQGASVRTFTPFYFLVEPVDTLSVRGSSVILNCSA...</td>\n",
       "      <td>7.0</td>\n",
       "      <td>doi.org/10.1038/s41592-020-0801-4</td>\n",
       "      <td>49.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31385</th>\n",
       "      <td>31385</td>\n",
       "      <td>YYMYSGGGSALAAGGGGAGRKGDWNDIDSIKKKDLHHSRGDEKAQG...</td>\n",
       "      <td>7.0</td>\n",
       "      <td>doi.org/10.1038/s41592-020-0801-4</td>\n",
       "      <td>51.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31386</th>\n",
       "      <td>31386</td>\n",
       "      <td>YYNDQHRLSSYSVETAMFLSWERAIVKPGAMFKKAVIGFNCNVDLI...</td>\n",
       "      <td>7.0</td>\n",
       "      <td>doi.org/10.1038/s41592-020-0801-4</td>\n",
       "      <td>37.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31387</th>\n",
       "      <td>31387</td>\n",
       "      <td>YYQRTLGAELLYKISFGEMPKSAQDSAENCPSGMQFPDTAIAHANV...</td>\n",
       "      <td>7.0</td>\n",
       "      <td>doi.org/10.1038/s41592-020-0801-4</td>\n",
       "      <td>64.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31388</th>\n",
       "      <td>31388</td>\n",
       "      <td>YYSFSDNITTVFLSRQAIDDDHSLSLGTISDVVESENGVVAADDAR...</td>\n",
       "      <td>7.0</td>\n",
       "      <td>doi.org/10.1038/s41592-020-0801-4</td>\n",
       "      <td>50.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31389</th>\n",
       "      <td>31389</td>\n",
       "      <td>YYVPDEYWQSLEVAHKLTFGYGYLTWEWVQGIRSYVYPLLIAGLYK...</td>\n",
       "      <td>7.0</td>\n",
       "      <td>doi.org/10.1038/s41592-020-0801-4</td>\n",
       "      <td>37.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>31390 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seq_id                                   protein_sequence   pH  \\\n",
       "0           0  AAAAKAAALALLGEAPEVVDIWLPAGWRQPFRVFRLERKGDGVLVG...  7.0   \n",
       "1           1  AAADGEPLHNEEERAGAGQVGRSLPQESEEQRTGSRPRRRRDLGSR...  7.0   \n",
       "2           2  AAAFSTPRATSYRILSSAGSGSTRADAPQVRRLHTTRDLLAKDYYA...  7.0   \n",
       "3           3  AAASGLRTAIPAQPLRHLLQPAPRPCLRPFGLLSVRAGSARRSGLL...  7.0   \n",
       "4           4  AAATKSGPRRQSQGASVRTFTPFYFLVEPVDTLSVRGSSVILNCSA...  7.0   \n",
       "...       ...                                                ...  ...   \n",
       "31385   31385  YYMYSGGGSALAAGGGGAGRKGDWNDIDSIKKKDLHHSRGDEKAQG...  7.0   \n",
       "31386   31386  YYNDQHRLSSYSVETAMFLSWERAIVKPGAMFKKAVIGFNCNVDLI...  7.0   \n",
       "31387   31387  YYQRTLGAELLYKISFGEMPKSAQDSAENCPSGMQFPDTAIAHANV...  7.0   \n",
       "31388   31388  YYSFSDNITTVFLSRQAIDDDHSLSLGTISDVVESENGVVAADDAR...  7.0   \n",
       "31389   31389  YYVPDEYWQSLEVAHKLTFGYGYLTWEWVQGIRSYVYPLLIAGLYK...  7.0   \n",
       "\n",
       "                             data_source    tm  \n",
       "0      doi.org/10.1038/s41592-020-0801-4  75.7  \n",
       "1      doi.org/10.1038/s41592-020-0801-4  50.5  \n",
       "2      doi.org/10.1038/s41592-020-0801-4  40.5  \n",
       "3      doi.org/10.1038/s41592-020-0801-4  47.2  \n",
       "4      doi.org/10.1038/s41592-020-0801-4  49.5  \n",
       "...                                  ...   ...  \n",
       "31385  doi.org/10.1038/s41592-020-0801-4  51.8  \n",
       "31386  doi.org/10.1038/s41592-020-0801-4  37.2  \n",
       "31387  doi.org/10.1038/s41592-020-0801-4  64.6  \n",
       "31388  doi.org/10.1038/s41592-020-0801-4  50.7  \n",
       "31389  doi.org/10.1038/s41592-020-0801-4  37.6  \n",
       "\n",
       "[31390 rows x 5 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>protein_sequence</th>\n",
       "      <th>pH</th>\n",
       "      <th>data_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>31390</td>\n",
       "      <td>VPVNPEPDATSVENVAEKTGSGDSQSDPIKADLEVKGQSALPFDVD...</td>\n",
       "      <td>8</td>\n",
       "      <td>Novozymes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>31391</td>\n",
       "      <td>VPVNPEPDATSVENVAKKTGSGDSQSDPIKADLEVKGQSALPFDVD...</td>\n",
       "      <td>8</td>\n",
       "      <td>Novozymes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>31392</td>\n",
       "      <td>VPVNPEPDATSVENVAKTGSGDSQSDPIKADLEVKGQSALPFDVDC...</td>\n",
       "      <td>8</td>\n",
       "      <td>Novozymes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>31393</td>\n",
       "      <td>VPVNPEPDATSVENVALCTGSGDSQSDPIKADLEVKGQSALPFDVD...</td>\n",
       "      <td>8</td>\n",
       "      <td>Novozymes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>31394</td>\n",
       "      <td>VPVNPEPDATSVENVALFTGSGDSQSDPIKADLEVKGQSALPFDVD...</td>\n",
       "      <td>8</td>\n",
       "      <td>Novozymes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2408</th>\n",
       "      <td>33798</td>\n",
       "      <td>VPVNPEPDATSVENVILKTGSGDSQSDPIKADLEVKGQSALPFDVD...</td>\n",
       "      <td>8</td>\n",
       "      <td>Novozymes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2409</th>\n",
       "      <td>33799</td>\n",
       "      <td>VPVNPEPDATSVENVLLKTGSGDSQSDPIKADLEVKGQSALPFDVD...</td>\n",
       "      <td>8</td>\n",
       "      <td>Novozymes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2410</th>\n",
       "      <td>33800</td>\n",
       "      <td>VPVNPEPDATSVENVNLKTGSGDSQSDPIKADLEVKGQSALPFDVD...</td>\n",
       "      <td>8</td>\n",
       "      <td>Novozymes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2411</th>\n",
       "      <td>33801</td>\n",
       "      <td>VPVNPEPDATSVENVPLKTGSGDSQSDPIKADLEVKGQSALPFDVD...</td>\n",
       "      <td>8</td>\n",
       "      <td>Novozymes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2412</th>\n",
       "      <td>33802</td>\n",
       "      <td>VPVNPEPDATSVENVWLKTGSGDSQSDPIKADLEVKGQSALPFDVD...</td>\n",
       "      <td>8</td>\n",
       "      <td>Novozymes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2413 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      seq_id                                   protein_sequence  pH  \\\n",
       "0      31390  VPVNPEPDATSVENVAEKTGSGDSQSDPIKADLEVKGQSALPFDVD...   8   \n",
       "1      31391  VPVNPEPDATSVENVAKKTGSGDSQSDPIKADLEVKGQSALPFDVD...   8   \n",
       "2      31392  VPVNPEPDATSVENVAKTGSGDSQSDPIKADLEVKGQSALPFDVDC...   8   \n",
       "3      31393  VPVNPEPDATSVENVALCTGSGDSQSDPIKADLEVKGQSALPFDVD...   8   \n",
       "4      31394  VPVNPEPDATSVENVALFTGSGDSQSDPIKADLEVKGQSALPFDVD...   8   \n",
       "...      ...                                                ...  ..   \n",
       "2408   33798  VPVNPEPDATSVENVILKTGSGDSQSDPIKADLEVKGQSALPFDVD...   8   \n",
       "2409   33799  VPVNPEPDATSVENVLLKTGSGDSQSDPIKADLEVKGQSALPFDVD...   8   \n",
       "2410   33800  VPVNPEPDATSVENVNLKTGSGDSQSDPIKADLEVKGQSALPFDVD...   8   \n",
       "2411   33801  VPVNPEPDATSVENVPLKTGSGDSQSDPIKADLEVKGQSALPFDVD...   8   \n",
       "2412   33802  VPVNPEPDATSVENVWLKTGSGDSQSDPIKADLEVKGQSALPFDVD...   8   \n",
       "\n",
       "     data_source  \n",
       "0      Novozymes  \n",
       "1      Novozymes  \n",
       "2      Novozymes  \n",
       "3      Novozymes  \n",
       "4      Novozymes  \n",
       "...          ...  \n",
       "2408   Novozymes  \n",
       "2409   Novozymes  \n",
       "2410   Novozymes  \n",
       "2411   Novozymes  \n",
       "2412   Novozymes  \n",
       "\n",
       "[2413 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>tm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>31390</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>31391</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>31392</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>31393</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>31394</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2408</th>\n",
       "      <td>33798</td>\n",
       "      <td>2408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2409</th>\n",
       "      <td>33799</td>\n",
       "      <td>2409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2410</th>\n",
       "      <td>33800</td>\n",
       "      <td>2410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2411</th>\n",
       "      <td>33801</td>\n",
       "      <td>2411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2412</th>\n",
       "      <td>33802</td>\n",
       "      <td>2412</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2413 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      seq_id    tm\n",
       "0      31390     0\n",
       "1      31391     1\n",
       "2      31392     2\n",
       "3      31393     3\n",
       "4      31394     4\n",
       "...      ...   ...\n",
       "2408   33798  2408\n",
       "2409   33799  2409\n",
       "2410   33800  2410\n",
       "2411   33801  2411\n",
       "2412   33802  2412\n",
       "\n",
       "[2413 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>record_name</th>\n",
       "      <th>atom_number</th>\n",
       "      <th>blank_1</th>\n",
       "      <th>atom_name</th>\n",
       "      <th>alt_loc</th>\n",
       "      <th>residue_name</th>\n",
       "      <th>blank_2</th>\n",
       "      <th>chain_id</th>\n",
       "      <th>residue_number</th>\n",
       "      <th>insertion</th>\n",
       "      <th>blank_3</th>\n",
       "      <th>x_coord</th>\n",
       "      <th>y_coord</th>\n",
       "      <th>z_coord</th>\n",
       "      <th>occupancy</th>\n",
       "      <th>b_factor</th>\n",
       "      <th>blank_4</th>\n",
       "      <th>segment_id</th>\n",
       "      <th>element_symbol</th>\n",
       "      <th>charge</th>\n",
       "      <th>line_idx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>N</td>\n",
       "      <td></td>\n",
       "      <td>VAL</td>\n",
       "      <td></td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>34.064</td>\n",
       "      <td>-6.456</td>\n",
       "      <td>50.464</td>\n",
       "      <td>1.0</td>\n",
       "      <td>45.11</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "      <td>H</td>\n",
       "      <td></td>\n",
       "      <td>VAL</td>\n",
       "      <td></td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>33.576</td>\n",
       "      <td>-6.009</td>\n",
       "      <td>51.228</td>\n",
       "      <td>1.0</td>\n",
       "      <td>45.11</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>H</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>3</td>\n",
       "      <td></td>\n",
       "      <td>H2</td>\n",
       "      <td></td>\n",
       "      <td>VAL</td>\n",
       "      <td></td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>33.882</td>\n",
       "      <td>-7.449</td>\n",
       "      <td>50.477</td>\n",
       "      <td>1.0</td>\n",
       "      <td>45.11</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>H</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>4</td>\n",
       "      <td></td>\n",
       "      <td>H3</td>\n",
       "      <td></td>\n",
       "      <td>VAL</td>\n",
       "      <td></td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>35.060</td>\n",
       "      <td>-6.323</td>\n",
       "      <td>50.566</td>\n",
       "      <td>1.0</td>\n",
       "      <td>45.11</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>H</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td>CA</td>\n",
       "      <td></td>\n",
       "      <td>VAL</td>\n",
       "      <td></td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>33.643</td>\n",
       "      <td>-5.877</td>\n",
       "      <td>49.162</td>\n",
       "      <td>1.0</td>\n",
       "      <td>45.11</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3312</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>3313</td>\n",
       "      <td></td>\n",
       "      <td>NZ</td>\n",
       "      <td></td>\n",
       "      <td>LYS</td>\n",
       "      <td></td>\n",
       "      <td>A</td>\n",
       "      <td>221</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4.616</td>\n",
       "      <td>13.323</td>\n",
       "      <td>-4.301</td>\n",
       "      <td>1.0</td>\n",
       "      <td>93.80</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3313</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>3314</td>\n",
       "      <td></td>\n",
       "      <td>HZ1</td>\n",
       "      <td></td>\n",
       "      <td>LYS</td>\n",
       "      <td></td>\n",
       "      <td>A</td>\n",
       "      <td>221</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>5.270</td>\n",
       "      <td>12.565</td>\n",
       "      <td>-4.432</td>\n",
       "      <td>1.0</td>\n",
       "      <td>93.80</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>H</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3314</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>3315</td>\n",
       "      <td></td>\n",
       "      <td>HZ2</td>\n",
       "      <td></td>\n",
       "      <td>LYS</td>\n",
       "      <td></td>\n",
       "      <td>A</td>\n",
       "      <td>221</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4.585</td>\n",
       "      <td>13.517</td>\n",
       "      <td>-3.310</td>\n",
       "      <td>1.0</td>\n",
       "      <td>93.80</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>H</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3315</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>3316</td>\n",
       "      <td></td>\n",
       "      <td>HZ3</td>\n",
       "      <td></td>\n",
       "      <td>LYS</td>\n",
       "      <td></td>\n",
       "      <td>A</td>\n",
       "      <td>221</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4.965</td>\n",
       "      <td>14.143</td>\n",
       "      <td>-4.776</td>\n",
       "      <td>1.0</td>\n",
       "      <td>93.80</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>H</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3316</th>\n",
       "      <td>ATOM</td>\n",
       "      <td>3317</td>\n",
       "      <td></td>\n",
       "      <td>OXT</td>\n",
       "      <td></td>\n",
       "      <td>LYS</td>\n",
       "      <td></td>\n",
       "      <td>A</td>\n",
       "      <td>221</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.511</td>\n",
       "      <td>13.763</td>\n",
       "      <td>-11.203</td>\n",
       "      <td>1.0</td>\n",
       "      <td>93.80</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>O</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3316</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3317 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     record_name  atom_number blank_1 atom_name alt_loc residue_name blank_2  \\\n",
       "0           ATOM            1                 N                  VAL           \n",
       "1           ATOM            2                 H                  VAL           \n",
       "2           ATOM            3                H2                  VAL           \n",
       "3           ATOM            4                H3                  VAL           \n",
       "4           ATOM            5                CA                  VAL           \n",
       "...          ...          ...     ...       ...     ...          ...     ...   \n",
       "3312        ATOM         3313                NZ                  LYS           \n",
       "3313        ATOM         3314               HZ1                  LYS           \n",
       "3314        ATOM         3315               HZ2                  LYS           \n",
       "3315        ATOM         3316               HZ3                  LYS           \n",
       "3316        ATOM         3317               OXT                  LYS           \n",
       "\n",
       "     chain_id  residue_number insertion blank_3  x_coord  y_coord  z_coord  \\\n",
       "0           A               1                     34.064   -6.456   50.464   \n",
       "1           A               1                     33.576   -6.009   51.228   \n",
       "2           A               1                     33.882   -7.449   50.477   \n",
       "3           A               1                     35.060   -6.323   50.566   \n",
       "4           A               1                     33.643   -5.877   49.162   \n",
       "...       ...             ...       ...     ...      ...      ...      ...   \n",
       "3312        A             221                      4.616   13.323   -4.301   \n",
       "3313        A             221                      5.270   12.565   -4.432   \n",
       "3314        A             221                      4.585   13.517   -3.310   \n",
       "3315        A             221                      4.965   14.143   -4.776   \n",
       "3316        A             221                      1.511   13.763  -11.203   \n",
       "\n",
       "      occupancy  b_factor blank_4 segment_id element_symbol  charge  line_idx  \n",
       "0           1.0     45.11                                 N     NaN         0  \n",
       "1           1.0     45.11                                 H     NaN         1  \n",
       "2           1.0     45.11                                 H     NaN         2  \n",
       "3           1.0     45.11                                 H     NaN         3  \n",
       "4           1.0     45.11                                 C     NaN         4  \n",
       "...         ...       ...     ...        ...            ...     ...       ...  \n",
       "3312        1.0     93.80                                 N     NaN      3312  \n",
       "3313        1.0     93.80                                 H     NaN      3313  \n",
       "3314        1.0     93.80                                 H     NaN      3314  \n",
       "3315        1.0     93.80                                 H     NaN      3315  \n",
       "3316        1.0     93.80                                 O     NaN      3316  \n",
       "\n",
       "[3317 rows x 21 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>record_name</th>\n",
       "      <th>atom_number</th>\n",
       "      <th>blank_1</th>\n",
       "      <th>atom_name</th>\n",
       "      <th>alt_loc</th>\n",
       "      <th>residue_name</th>\n",
       "      <th>blank_2</th>\n",
       "      <th>chain_id</th>\n",
       "      <th>residue_number</th>\n",
       "      <th>insertion</th>\n",
       "      <th>blank_3</th>\n",
       "      <th>x_coord</th>\n",
       "      <th>y_coord</th>\n",
       "      <th>z_coord</th>\n",
       "      <th>occupancy</th>\n",
       "      <th>b_factor</th>\n",
       "      <th>blank_4</th>\n",
       "      <th>segment_id</th>\n",
       "      <th>element_symbol</th>\n",
       "      <th>charge</th>\n",
       "      <th>line_idx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [record_name, atom_number, blank_1, atom_name, alt_loc, residue_name, blank_2, chain_id, residue_number, insertion, blank_3, x_coord, y_coord, z_coord, occupancy, b_factor, blank_4, segment_id, element_symbol, charge, line_idx]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>record_name</th>\n",
       "      <th>atom_number</th>\n",
       "      <th>blank_1</th>\n",
       "      <th>atom_name</th>\n",
       "      <th>alt_loc</th>\n",
       "      <th>residue_name</th>\n",
       "      <th>blank_2</th>\n",
       "      <th>chain_id</th>\n",
       "      <th>residue_number</th>\n",
       "      <th>insertion</th>\n",
       "      <th>blank_3</th>\n",
       "      <th>U(1,1)</th>\n",
       "      <th>U(2,2)</th>\n",
       "      <th>U(3,3)</th>\n",
       "      <th>U(1,2)</th>\n",
       "      <th>U(1,3)</th>\n",
       "      <th>U(2,3)</th>\n",
       "      <th>blank_4</th>\n",
       "      <th>element_symbol</th>\n",
       "      <th>charge</th>\n",
       "      <th>line_idx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [record_name, atom_number, blank_1, atom_name, alt_loc, residue_name, blank_2, chain_id, residue_number, insertion, blank_3, U(1,1), U(2,2), U(3,3), U(1,2), U(1,3), U(2,3), blank_4, element_symbol, charge, line_idx]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>record_name</th>\n",
       "      <th>entry</th>\n",
       "      <th>line_idx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TER</td>\n",
       "      <td>3318      LYS A 221</td>\n",
       "      <td>3317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>END</td>\n",
       "      <td></td>\n",
       "      <td>3318</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  record_name                 entry  line_idx\n",
       "0         TER   3318      LYS A 221      3317\n",
       "1         END                            3318"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VPVNPEPDATSVENVALKTGSGDSQSDPIKADLEVKGQSALPFDVDCWAILCKGAPNVLQRVNEKTKNSNRDRSGANKGPFKDPQKWGIKALPPKNPSWSAQDFKSPEEYAFASSLQGGTNAILAPVNLASQNSQGGVLNGFYSANKVAQFDPSKPQQTKGTWFQITKFTGAAGPYCKALGSNDKSVCDKNKNIAGDWGFDPAKWAYQYDEKNNKFNYVGK\n",
      "'Alanine':\n",
      "\t3 LETTER ABBREVIATION --> 'Ala'\n",
      "\t1 LETTER ABBREVIATION --> 'A'\n",
      "\n",
      "'Arginine':\n",
      "\t3 LETTER ABBREVIATION --> 'Arg'\n",
      "\t1 LETTER ABBREVIATION --> 'R'\n",
      "\n",
      "'Asparagine':\n",
      "\t3 LETTER ABBREVIATION --> 'Asn'\n",
      "\t1 LETTER ABBREVIATION --> 'N'\n",
      "\n",
      "'Aspartic_Acid':\n",
      "\t3 LETTER ABBREVIATION --> 'Asp'\n",
      "\t1 LETTER ABBREVIATION --> 'D'\n",
      "\n",
      "'Cysteine':\n",
      "\t3 LETTER ABBREVIATION --> 'Cys'\n",
      "\t1 LETTER ABBREVIATION --> 'C'\n",
      "\n",
      "'Glutamic_Acid':\n",
      "\t3 LETTER ABBREVIATION --> 'Glu'\n",
      "\t1 LETTER ABBREVIATION --> 'E'\n",
      "\n",
      "'Glutamine':\n",
      "\t3 LETTER ABBREVIATION --> 'Gln'\n",
      "\t1 LETTER ABBREVIATION --> 'Q'\n",
      "\n",
      "'Glycine':\n",
      "\t3 LETTER ABBREVIATION --> 'Gly'\n",
      "\t1 LETTER ABBREVIATION --> 'G'\n",
      "\n",
      "'Histidine':\n",
      "\t3 LETTER ABBREVIATION --> 'His'\n",
      "\t1 LETTER ABBREVIATION --> 'H'\n",
      "\n",
      "'Isoleucine':\n",
      "\t3 LETTER ABBREVIATION --> 'Ile'\n",
      "\t1 LETTER ABBREVIATION --> 'I'\n",
      "\n",
      "'Leucine':\n",
      "\t3 LETTER ABBREVIATION --> 'Leu'\n",
      "\t1 LETTER ABBREVIATION --> 'L'\n",
      "\n",
      "'Lysine':\n",
      "\t3 LETTER ABBREVIATION --> 'Lys'\n",
      "\t1 LETTER ABBREVIATION --> 'K'\n",
      "\n",
      "'Methionine':\n",
      "\t3 LETTER ABBREVIATION --> 'Met'\n",
      "\t1 LETTER ABBREVIATION --> 'M'\n",
      "\n",
      "'Phenylalanine':\n",
      "\t3 LETTER ABBREVIATION --> 'Phe'\n",
      "\t1 LETTER ABBREVIATION --> 'F'\n",
      "\n",
      "'Proline':\n",
      "\t3 LETTER ABBREVIATION --> 'Pro'\n",
      "\t1 LETTER ABBREVIATION --> 'P'\n",
      "\n",
      "'Serine':\n",
      "\t3 LETTER ABBREVIATION --> 'Ser'\n",
      "\t1 LETTER ABBREVIATION --> 'S'\n",
      "\n",
      "'Threonine':\n",
      "\t3 LETTER ABBREVIATION --> 'Thr'\n",
      "\t1 LETTER ABBREVIATION --> 'T'\n",
      "\n",
      "'Tryptophan':\n",
      "\t3 LETTER ABBREVIATION --> 'Trp'\n",
      "\t1 LETTER ABBREVIATION --> 'W'\n",
      "\n",
      "'Tyrosine':\n",
      "\t3 LETTER ABBREVIATION --> 'Tyr'\n",
      "\t1 LETTER ABBREVIATION --> 'Y'\n",
      "\n",
      "'Valine':\n",
      "\t3 LETTER ABBREVIATION --> 'Val'\n",
      "\t1 LETTER ABBREVIATION --> 'V'\n",
      "\n",
      "\n",
      "\n",
      "... FOR FUN ... HERE IS THE ENTIRE WILDTYPE WITH FULL AMINO ACID NAMES (8 AA PER LINE) ...\n",
      "\n",
      "'Valine', 'Proline', 'Valine', 'Asparagine', 'Proline', 'Glutamic_Acid', 'Proline', Aspartic_Acid,\n",
      "'Alanine', 'Threonine', 'Serine', 'Valine', 'Glutamic_Acid', 'Asparagine', 'Valine', Alanine,\n",
      "'Leucine', 'Lysine', 'Threonine', 'Glycine', 'Serine', 'Glycine', 'Aspartic_Acid', Serine,\n",
      "'Glutamine', 'Serine', 'Aspartic_Acid', 'Proline', 'Isoleucine', 'Lysine', 'Alanine', Aspartic_Acid,\n",
      "'Leucine', 'Glutamic_Acid', 'Valine', 'Lysine', 'Glycine', 'Glutamine', 'Serine', Alanine,\n",
      "'Leucine', 'Proline', 'Phenylalanine', 'Aspartic_Acid', 'Valine', 'Aspartic_Acid', 'Cysteine', Tryptophan,\n",
      "'Alanine', 'Isoleucine', 'Leucine', 'Cysteine', 'Lysine', 'Glycine', 'Alanine', Proline,\n",
      "'Asparagine', 'Valine', 'Leucine', 'Glutamine', 'Arginine', 'Valine', 'Asparagine', Glutamic_Acid,\n",
      "'Lysine', 'Threonine', 'Lysine', 'Asparagine', 'Serine', 'Asparagine', 'Arginine', Aspartic_Acid,\n",
      "'Arginine', 'Serine', 'Glycine', 'Alanine', 'Asparagine', 'Lysine', 'Glycine', Proline,\n",
      "'Phenylalanine', 'Lysine', 'Aspartic_Acid', 'Proline', 'Glutamine', 'Lysine', 'Tryptophan', Glycine,\n",
      "'Isoleucine', 'Lysine', 'Alanine', 'Leucine', 'Proline', 'Proline', 'Lysine', Asparagine,\n",
      "'Proline', 'Serine', 'Tryptophan', 'Serine', 'Alanine', 'Glutamine', 'Aspartic_Acid', Phenylalanine,\n",
      "'Lysine', 'Serine', 'Proline', 'Glutamic_Acid', 'Glutamic_Acid', 'Tyrosine', 'Alanine', Phenylalanine,\n",
      "'Alanine', 'Serine', 'Serine', 'Leucine', 'Glutamine', 'Glycine', 'Glycine', Threonine,\n",
      "'Asparagine', 'Alanine', 'Isoleucine', 'Leucine', 'Alanine', 'Proline', 'Valine', Asparagine,\n",
      "'Leucine', 'Alanine', 'Serine', 'Glutamine', 'Asparagine', 'Serine', 'Glutamine', Glycine,\n",
      "'Glycine', 'Valine', 'Leucine', 'Asparagine', 'Glycine', 'Phenylalanine', 'Tyrosine', Serine,\n",
      "'Alanine', 'Asparagine', 'Lysine', 'Valine', 'Alanine', 'Glutamine', 'Phenylalanine', Aspartic_Acid,\n",
      "'Proline', 'Serine', 'Lysine', 'Proline', 'Glutamine', 'Glutamine', 'Threonine', Lysine,\n",
      "'Glycine', 'Threonine', 'Tryptophan', 'Phenylalanine', 'Glutamine', 'Isoleucine', 'Threonine', Lysine,\n",
      "'Phenylalanine', 'Threonine', 'Glycine', 'Alanine', 'Alanine', 'Glycine', 'Proline', Tyrosine,\n",
      "'Cysteine', 'Lysine', 'Alanine', 'Leucine', 'Glycine', 'Serine', 'Asparagine', Aspartic_Acid,\n",
      "'Lysine', 'Serine', 'Valine', 'Cysteine', 'Aspartic_Acid', 'Lysine', 'Asparagine', Lysine,\n",
      "'Asparagine', 'Isoleucine', 'Alanine', 'Glycine', 'Aspartic_Acid', 'Tryptophan', 'Glycine', Phenylalanine,\n",
      "'Aspartic_Acid', 'Proline', 'Alanine', 'Lysine', 'Tryptophan', 'Alanine', 'Tyrosine', Glutamine,\n",
      "'Tyrosine', 'Aspartic_Acid', 'Glutamic_Acid', 'Lysine', 'Asparagine', 'Asparagine', 'Lysine', Phenylalanine,\n",
      "'Asparagine', 'Tyrosine', 'Valine', 'Glycine', 'Lysine', \n",
      "\n",
      "... ADD COLUMNS FOR PROTEIN SEQUENCE LENGTH AND INDIVIDUAL AMINO ACID COUNTS/FRACTIONS ...\n",
      "\n",
      "\n",
      "... ADD COLUMNS FOR DATA SOURCE ENUMERATION ...\n",
      "\n",
      "\n",
      "... DO TEMPORARY pH FIX BY SWAPPING pH & tm IF pH>14 ...\n",
      "\n",
      "\t--> DOES THE  STILL EXIST: True\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8a7224cf9acd44a18cc61a745fd7351f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/31390 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t--> pH WEIRDNESS EXISTS AT INDEX 973\n",
      "\t--> pH WEIRDNESS EXISTS AT INDEX 986\n",
      "\t--> pH WEIRDNESS EXISTS AT INDEX 988\n",
      "\t--> pH WEIRDNESS EXISTS AT INDEX 989\n",
      "\t--> pH WEIRDNESS EXISTS AT INDEX 1003\n",
      "\t--> pH WEIRDNESS EXISTS AT INDEX 1012\n",
      "\t--> pH WEIRDNESS EXISTS AT INDEX 1014\n",
      "\t--> pH WEIRDNESS EXISTS AT INDEX 1018\n",
      "\t--> pH WEIRDNESS EXISTS AT INDEX 1037\n",
      "\t--> pH WEIRDNESS EXISTS AT INDEX 1042\n",
      "\t--> pH WEIRDNESS EXISTS AT INDEX 1079\n",
      "\t--> pH WEIRDNESS EXISTS AT INDEX 1081\n",
      "\t--> pH WEIRDNESS EXISTS AT INDEX 1088\n",
      "\t--> pH WEIRDNESS EXISTS AT INDEX 1093\n",
      "\t--> pH WEIRDNESS EXISTS AT INDEX 1096\n",
      "\t--> pH WEIRDNESS EXISTS AT INDEX 1098\n",
      "\t--> pH WEIRDNESS EXISTS AT INDEX 1100\n",
      "\t--> pH WEIRDNESS EXISTS AT INDEX 1108\n",
      "\t--> pH WEIRDNESS EXISTS AT INDEX 1111\n",
      "\t--> pH WEIRDNESS EXISTS AT INDEX 1120\n",
      "\t--> pH WEIRDNESS EXISTS AT INDEX 1122\n",
      "\t--> pH WEIRDNESS EXISTS AT INDEX 1125\n",
      "\t--> pH WEIRDNESS EXISTS AT INDEX 13447\n",
      "\t--> pH WEIRDNESS EXISTS AT INDEX 13449\n",
      "\t--> pH WEIRDNESS EXISTS AT INDEX 14640\n",
      "\t--> pH WEIRDNESS EXISTS AT INDEX 22116\n",
      "\t--> pH WEIRDNESS EXISTS AT INDEX 22123\n",
      "\t--> pH WEIRDNESS EXISTS AT INDEX 22131\n",
      "\t--> pH WEIRDNESS EXISTS AT INDEX 22145\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ec3e4980073540299f86d5463800838d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/2413 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Define the path to the root data directory\n",
    "DATA_DIR = \"/home/linux/Workspace/mutating-doodle\"\n",
    "train_df = pd.read_csv(os.path.join(DATA_DIR, \"train.csv\"))\n",
    "display(train_df)\n",
    "\n",
    "# loading test dataframe from csv file\n",
    "test_df = pd.read_csv(os.path.join(DATA_DIR, \"test.csv\"))\n",
    "display(test_df)\n",
    "\n",
    "# loading sample submission dataframe from csv file\n",
    "ss_df = pd.read_csv(os.path.join(DATA_DIR, \"sample_submission.csv\"))\n",
    "display(ss_df)\n",
    "\n",
    "# load alphafold wildtype structure data from pdb file\n",
    "pdb_df = pdb.read_pdb(os.path.join(DATA_DIR, \"wildtype_structure_prediction_af2.pdb\"))\n",
    "\n",
    "# atom data\n",
    "atom_df = pdb_df.df['ATOM']\n",
    "display(atom_df)\n",
    "\n",
    "# hetatm data\n",
    "hetatm_df = pdb_df.df['HETATM']\n",
    "display(hetatm_df)\n",
    "\n",
    "# anisou data\n",
    "anisou_df = pdb_df.df['ANISOU']\n",
    "display(anisou_df)\n",
    "\n",
    "# other data\n",
    "others_df = pdb_df.df['OTHERS']\n",
    "display(others_df)\n",
    "\n",
    "# saving wildtype amino acid sequence\n",
    "wildtype_aa = \"VPVNPEPDATSVENVALKTGSGDSQSDPIKADLEVKGQSALPFDVDCWAILCKGAPNVLQRVNEKTKNSNRDRSGANKGPFKDPQKWGIKALPPKNPSWSAQDFKSPEEYAFASSLQGGTNAILAPVNLASQNSQGGVLNGFYSANKVAQFDPSKPQQTKGTWFQITKFTGAAGPYCKALGSNDKSVCDKNKNIAGDWGFDPAKWAYQYDEKNNKFNYVGK\"\n",
    "print(wildtype_aa)\n",
    "\n",
    "# define amino acid shortform dictionary mapping\n",
    "aa_map = dict(Alanine=(\"Ala\", \"A\"), Arginine=(\"Arg\", \"R\"), Asparagine=(\"Asn\", \"N\"), Aspartic_Acid=(\"Asp\", \"D\"),\n",
    "              Cysteine=(\"Cys\", \"C\"), Glutamic_Acid=(\"Glu\", \"E\"), Glutamine=(\"Gln\", \"Q\"), Glycine=(\"Gly\", \"G\"),\n",
    "              Histidine=(\"His\", \"H\"), Isoleucine=(\"Ile\", \"I\"), Leucine=(\"Leu\", \"L\"), Lysine=(\"Lys\", \"K\"),\n",
    "              Methionine=(\"Met\", \"M\"), Phenylalanine=(\"Phe\", \"F\"), Proline=(\"Pro\", \"P\"), Serine=(\"Ser\", \"S\"),\n",
    "              Threonine=(\"Thr\", \"T\"), Tryptophan=(\"Trp\", \"W\"), Tyrosine=(\"Tyr\", \"Y\"), Valine=(\"Val\", \"V\"))\n",
    "n_aa = len(aa_map)\n",
    "aa_chars_ordered = sorted([v[1] for v in aa_map.values()])\n",
    "aa_long2tri = {k:v[0] for k,v in aa_map.items()}\n",
    "aa_long2char = {k:v[1] for k,v in aa_map.items()}\n",
    "aa_tri2long = {v:k for k,v in aa_long2tri.items()}\n",
    "aa_char2long = {v:k for k,v in aa_long2char.items()}\n",
    "aa_char2int = {_aa:i for i, _aa in enumerate(aa_chars_ordered)}\n",
    "aa_int2char = {v:k for k,v in aa_char2int.items()}\n",
    "\n",
    "# Get data source map\n",
    "ds_str2int = {k:i for i,k in enumerate(train_df[\"data_source\"].unique())}\n",
    "ds_int2str = {v:k for k,v in ds_str2int.items()}\n",
    "\n",
    "for k,v in aa_map.items(): print(f\"'{k}':\\n\\t3 LETTER ABBREVIATION --> '{v[0]}'\\n\\t1 LETTER ABBREVIATION --> '{v[1]}'\\n\")\n",
    "    \n",
    "print(\"\\n\\n... FOR FUN ... HERE IS THE ENTIRE WILDTYPE WITH FULL AMINO ACID NAMES (8 AA PER LINE) ...\\n\")\n",
    "for i, _c in enumerate(wildtype_aa): print(f\"'{aa_char2long[_c]}'\", end=\", \") if (i+1)%8!=0 else print(f\"{aa_char2long[_c]}\", end=\",\\n\")\n",
    "\n",
    "print(\"\\n\\n... ADD COLUMNS FOR PROTEIN SEQUENCE LENGTH AND INDIVIDUAL AMINO ACID COUNTS/FRACTIONS ...\\n\")\n",
    "train_df[\"n_AA\"] = train_df[\"protein_sequence\"].apply(len)\n",
    "test_df[\"n_AA\"] = test_df[\"protein_sequence\"].apply(len)\n",
    "for _aa_char in aa_chars_ordered: \n",
    "    train_df[f\"AA_{_aa_char}__count\"] = train_df[\"protein_sequence\"].apply(lambda x: x.count(_aa_char))\n",
    "    train_df[f\"AA_{_aa_char}__fraction\"] = train_df[f\"AA_{_aa_char}__count\"]/train_df[\"n_AA\"]\n",
    "    test_df[f\"AA_{_aa_char}__count\"] = test_df[\"protein_sequence\"].apply(lambda x: x.count(_aa_char))\n",
    "    test_df[f\"AA_{_aa_char}__fraction\"] = test_df[f\"AA_{_aa_char}__count\"]/test_df[\"n_AA\"]\n",
    "    \n",
    "print(\"\\n... ADD COLUMNS FOR DATA SOURCE ENUMERATION ...\\n\")\n",
    "train_df[\"data_source_enum\"] = train_df['data_source'].map(ds_str2int)\n",
    "test_df[\"data_source_enum\"] = test_df['data_source'].map(ds_str2int)\n",
    "\n",
    "print(\"\\n... DO TEMPORARY pH FIX BY SWAPPING pH & tm IF pH>14 ...\\n\")\n",
    "def tmp_ph_fix(_row):\n",
    "    if _row[\"pH\"]>14:\n",
    "        print(f\"\\t--> pH WEIRDNESS EXISTS AT INDEX {_row.name}\")\n",
    "        _tmp = _row[\"pH\"]\n",
    "        _row[\"pH\"] = _row[\"tm\"]\n",
    "        _row[\"tm\"] = _tmp\n",
    "        return _row\n",
    "    else:\n",
    "        return _row\n",
    "\n",
    "print(f\"\\t--> DOES THE  STILL EXIST: {train_df['pH'].max()>14.0}\")\n",
    "train_df = train_df.progress_apply(tmp_ph_fix, axis=1)\n",
    "test_df = test_df.progress_apply(tmp_ph_fix, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1.1 – Define the path to the model output text file (I'm using mine but you can use the original)\n",
    "DEEPDDG_PRED_TXT_PATH = \"/kaggle/input/novoesp-deepddg-server-predictions-sub-only/wildtype_structure_prediction_af2.ddg.txt\"\n",
    "\n",
    "# 1.2 – Read the txt file into a pandas dataframe specifying a space as the delimiter\n",
    "#       --> This introduces some weirdness with the number of columns... so we have to drop some and rename them\n",
    "#       --> We also drop the #chain column as it contains no usable information (all rows are the same --> A)\n",
    "#       --> We also rename the columns to be more user friendly (note ddg stands for ΔΔG)\n",
    "deepddg_pred_df = pd.read_table(DEEPDDG_PRED_TXT_PATH, sep=\" \").drop(columns=[\"#chain\", \"ddG\", \"is\", \"stable,\", \"is.1\", \"unstable)\", \"<0\"])\n",
    "deepddg_pred_df.columns = [\"wildtype_aa\", \"residue_id\", \"mutant_aa\", \"ddg\", \"ddg_\"]\n",
    "\n",
    "# 1.3 – Coerce all the ddg values to be in the right column\n",
    "deepddg_pred_df[\"ddg\"] = deepddg_pred_df[\"ddg\"].fillna(0.0)+deepddg_pred_df[\"ddg_\"].fillna(0.0)\n",
    "deepddg_pred_df = deepddg_pred_df.drop(columns=[\"ddg_\"])\n",
    "\n",
    "# 1.4 – Change edit location string name and change from 1-indexed to 0-indexed\n",
    "deepddg_pred_df.loc[:,'location'] = deepddg_pred_df[\"residue_id\"]-1\n",
    "deepddg_pred_df = deepddg_pred_df.drop(columns=[\"residue_id\"])\n",
    "\n",
    "# 1.5 – Create a new column that contains the entire mutation as a string\n",
    "#   --> This mutation string has the format   <wildtype_aa><location><mutant_aa>\n",
    "deepddg_pred_df.loc[:,'mutant_string'] = deepddg_pred_df[\"wildtype_aa\"]+deepddg_pred_df[\"location\"].astype(str)+deepddg_pred_df[\"mutant_aa\"]\n",
    "\n",
    "# 1.6 – Display the newly created datafarme containing predictions (and describe float/int based columns)\n",
    "display(deepddg_pred_df.describe().T)\n",
    "display(deepddg_pred_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2.1 – Adjust the residue column to be 0-indexed instead of 1-indexed\n",
    "atom_df['residue_number'] -= 1\n",
    "\n",
    "# 2.2 – Create a mapping from residue to b-factor \n",
    "#   --> b-factor is always the same for a particular residue\n",
    "residue_to_bfactor_map = atom_df.groupby('residue_number').b_factor.first().to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 3.1 – Add mutation information about test set\n",
    "#   --> Type of Mutation ['substitution'|'deletion'|'no_change'|'insertion'] (insertion never occurs)\n",
    "#   --> Position of Mutation (Note this is now 0 indexed from the previous cell/section)\n",
    "#   --> Wildtype Amino Acid (Single Letter Short Form)\n",
    "#   --> Mutant Amino Acid (Single Letter Short Form – Will be NaN in 'deletion' mutations)\n",
    "test_df = test_df.progress_apply(get_mutation_info, axis=1)\n",
    "\n",
    "# 3.2 – Add b-factor to the test dataframe using previously created dictionary\n",
    "test_df['b_factor'] = test_df[\"edit_idx\"].map(residue_to_bfactor_map)\n",
    "\n",
    "# 3.3 – Change edit_type from NaN to 'no_change'\n",
    "#   --> this will allow the entire column to be a string as NaN is considered a float\n",
    "test_df[\"edit_type\"] = test_df[\"edit_type\"].fillna(\"no_change\")\n",
    "\n",
    "# 3.4 – Change mutant_aa from NaN to '+' or '-' if edit_type is 'insertion' or 'deletion' respectively \n",
    "#   --> this will allow the entire column to be a string as NaN is considered a float\n",
    "test_df.loc[test_df['edit_type']=='deletion', 'mutant_aa'] = '-'\n",
    "test_df.loc[test_df['edit_type']=='insertion', 'mutant_aa'] = '+'\n",
    "\n",
    "# 3.5 – Create a new column that contains the entire mutation as a string\n",
    "#   --> This mutation string has the format   <wildtype_aa><edit_idx><mutant_aa>\n",
    "test_df.loc[:, 'mutant_string'] =  test_df[\"wildtype_aa\"]+test_df[\"edit_idx\"].astype(str)+test_df[\"mutant_aa\"]\n",
    "test_df[[\"seq_id\", \"edit_type\", \"edit_idx\", \"wildtype_aa\", \"mutant_aa\", \"b_factor\",\"mutant_string\"]].head()\n",
    "\n",
    "# 3.6 – Drop columns we don't need for this notebook\n",
    "test_df = test_df[[_c for _c in test_df.columns if ((\"__\" not in _c) and (\"data_source\" not in _c))]]\n",
    "\n",
    "# 3.7 – Display the updated dataframe (and describe float/int based columns)\n",
    "display(test_df.describe().T)\n",
    "display(test_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4.1 – Merge the two dataframes together \n",
    "test_df = test_df.merge(deepddg_pred_df[['ddg','mutant_string']], on='mutant_string', how='left')\n",
    "\n",
    "# 4.2 – Fill in the missing ddg values with predetermined values\n",
    "#   --> We set the default value for deletion to be equivalent to the bottom quartile value\n",
    "#       of all substitutions... this is because it is more deleterious than simple substitutions\n",
    "#   --> The default no_change value is simply 0.0 because this is the wildtype\n",
    "#   --> What I would have thought is better: \n",
    "#            ----> DEFAULT__DELETION__DDG  = test_df[test_df[\"edit_type\"]==\"substitution\"][\"ddg\"].quantile(q=0.25)\n",
    "DEFAULT__DELETION__DDG  = -0.25\n",
    "DEFAULT__NO_CHANGE__DDG = 0.0   # THIS IS DIFFERENT THAN THE ORIGINAL NOTEBOOK\n",
    "test_df.loc[test_df['edit_type']==\"deletion\", 'ddg'] = DEFAULT__DELETION__DDG\n",
    "test_df.loc[test_df['edit_type']==\"no_change\", 'ddg'] = DEFAULT__NO_CHANGE__DDG\n",
    "\n",
    "# 4.3 – Display the updated dataframe (and describe float/int based columns)\n",
    "display(test_df.describe().T)\n",
    "display(test_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 5.1 – Define a function to return the substitution matrix (backwards and forwards)\n",
    "from Bio.SubsMat import MatrixInfo\n",
    "def get_sub_matrix(matrix_name=\"blosum100\"):\n",
    "    sub_matrix = getattr(MatrixInfo, matrix_name)\n",
    "    sub_matrix.update({(k[1], k[0]):v for k,v in sub_matrix.items() if (k[1], k[0]) not in list(sub_matrix.keys())})\n",
    "    return sub_matrix\n",
    "sub_matrix = get_sub_matrix()\n",
    "\n",
    "# 5.2 – Conduct matrix substitution\n",
    "#   --> First we create a tuple that has the wildtype amino acid and the \n",
    "#       mutant amino acid to access the substitution matrix\n",
    "#   --> Second we access the substitution matrix and replace with the respective score\n",
    "#       and in cases where no respective score is found we mark it to be updated later\n",
    "test_df[\"sub_matrix_tuple\"] = test_df[[\"wildtype_aa\", \"mutant_aa\"]].apply(tuple, axis=1)\n",
    "test_df[\"sub_matrix_score\"] = test_df[\"sub_matrix_tuple\"].progress_apply(lambda _mutant_tuple: sub_matrix.get(_mutant_tuple, \"tbd\"))\n",
    "\n",
    "# 5.3 – Fill in the missing data with default values for now\n",
    "#   --> We set the default value for matrix sub to be equivalent to the bottom quartile value\n",
    "#       of all substitutions... this is because it is more deleterious than simple substitutions (larger difference)\n",
    "#   --> The default no_change value is 1 higher than the max score because a higher score means more similarity\n",
    "#DEFAULT__DELETION__MATRIXSCORE  = test_df[test_df[\"edit_type\"]==\"substitution\"][\"sub_matrix_score\"].quantile(q=0.25)\n",
    "DEFAULT__DELETION__MATRIXSCORE = -10.0\n",
    "#DEFAULT__NO_CHANGE__MATRIXSCORE = test_df[test_df[\"edit_type\"]==\"substitution\"][\"sub_matrix_score\"].max()+1.0\n",
    "DEFAULT__NO_CHANGE__MATRIXSCORE = 0.0\n",
    "test_df.loc[test_df['edit_type']==\"deletion\", 'sub_matrix_score'] = DEFAULT__DELETION__MATRIXSCORE\n",
    "test_df.loc[test_df['edit_type']==\"no_change\", 'sub_matrix_score'] = DEFAULT__NO_CHANGE__MATRIXSCORE\n",
    "test_df[\"sub_matrix_score\"] = test_df[\"sub_matrix_score\"].astype(float)\n",
    "\n",
    "# 5.4 – Display the updated dataframe (and describe float/int based columns)\n",
    "display(test_df.describe().T)\n",
    "display(test_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6.1 – If the flag is set, reduce all positive matrix substitution scores to 0\n",
    "CAP_SUB_MATRIX_SCORE = True\n",
    "CAP_ABOVE_VAL, CAP_VAL = 0.0, 0.0\n",
    "if CAP_SUB_MATRIX_SCORE:\n",
    "    test_df.loc[test_df['sub_matrix_score'] > CAP_ABOVE_VAL, 'sub_matrix_score'] = CAP_VAL\n",
    "\n",
    "# 6.2 – Normalize the matrix substitution score with adjusted sigmoid function\n",
    "SIGMOID_ADJUSTMENT_CONSTANT = 3\n",
    "def sigmoid_w_adjustment(x, adjustment_factor=3.0):\n",
    "    return 1-(1/(1+np.exp(-x/adjustment_factor)))\n",
    "test_df['sub_matrix_score_normalized'] = test_df['sub_matrix_score'].apply(lambda x: sigmoid_w_adjustment(x, SIGMOID_ADJUSTMENT_CONSTANT))\n",
    "test_df['b_factor_matrix_score_adjusted'] = test_df['b_factor']*test_df['sub_matrix_score_normalized'] \n",
    "\n",
    "# 6.3 – Display the updated dataframe (and describe float/int based columns)\n",
    "display(test_df.describe().T)\n",
    "display(test_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy import stats\n",
    "\n",
    "# 7.1 – Assign ranks to data, dealing with ties appropriately.\n",
    "test_df['ddg_rank'] = stats.rankdata(test_df['ddg'])\n",
    "test_df['sub_matrix_score_rank'] = stats.rankdata(test_df['sub_matrix_score'])\n",
    "test_df['b_factor_rank'] = stats.rankdata(-test_df['b_factor'])\n",
    "test_df['b_factor_matrix_score_adjusted_rank'] = stats.rankdata(-test_df['b_factor_matrix_score_adjusted'])\n",
    "\n",
    "# 7.2 – Display the updated dataframe (and describe float/int based columns)\n",
    "display(test_df.describe().T)\n",
    "display(test_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 8.1 – Identify the columns to be combined\n",
    "combo_cols = ['b_factor_rank', 'sub_matrix_score_rank', 'ddg_rank']\n",
    "\n",
    "# 8.2 – Combine the columns by multiplying them together\n",
    "test_df[\"combined_val\"] = test_df[combo_cols].apply(np.prod, axis=1)\n",
    "\n",
    "# 8.3 – Plot the newly created column to see the distribution of values\n",
    "#       prior to any type of manipulation (i.e. raising to 1/3 power)\n",
    "plt.figure(figsize=(14,6))\n",
    "test_df[\"combined_val\"].hist()\n",
    "plt.title(\"Distribution of Values Prior To Being Raised to 1/3 Power\")\n",
    "plt.xlabel(\"b_factor_rank * sub_matrix_score_rank * ddg_rank\", fontweight=\"bold\")\n",
    "plt.ylabel(\"frequency of occurence in dataset\", fontweight=\"bold\")\n",
    "plt.show()\n",
    "\n",
    "# 8.4 – 'Normalize' the combined value column by raising to a particular power\n",
    "COMBO_NORM_PWR = 1/3\n",
    "test_df[\"norm_combined_val\"] = test_df[\"combined_val\"]**COMBO_NORM_PWR\n",
    "\n",
    "# 8.5 – Plot the newly created column to see the distribution of values\n",
    "#       after manipulation (i.e. raising to 1/3 power)\n",
    "plt.figure(figsize=(14,6))\n",
    "test_df[\"norm_combined_val\"].hist()\n",
    "plt.title(\"Distribution of Combined Normalized Values After Being Raised to 1/3 Power\")\n",
    "plt.xlabel(\"(b_factor_rank * sub_matrix_score_rank * ddg_rank)^(1/3)\", fontweight=\"bold\")\n",
    "plt.ylabel(\"frequency of occurence in dataset\", fontweight=\"bold\")\n",
    "plt.show()\n",
    "\n",
    "# 8.6 – Display the updated dataframe (and describe float/int based columns)\n",
    "display(test_df.describe().T)\n",
    "display(test_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 9.1 – Check that no NaN values slipped through\n",
    "assert not test_df[\"norm_combined_val\"].isna().any()\n",
    "\n",
    "# 9.2 – Update the sample submission dataframe by creating a mapping and\n",
    "#       applying it to the previously created sample_submission.csv\n",
    "seqid_2_tmrank_mapping = test_df.groupby(\"seq_id\")[\"norm_combined_val\"].first().to_dict()\n",
    "ss_df[\"tm\"] = ss_df[\"seq_id\"].apply(lambda x: seqid_2_tmrank_mapping[x])\n",
    "\n",
    "# 9.3 – Save properly for submission\n",
    "ss_df.to_csv(\"submission.csv\", index=False)\n",
    "\n",
    "# 9.4 – Display the sample submission dataframe and it's details\n",
    "display(ss_df.describe().T)\n",
    "display(ss_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1.1 – Create substitution matrix for each position from DeMaSk data (apologies for the dump file names)\n",
    "demask_sub_matrix = pd.read_csv(\"../input/demask-directional-substitution-matrix/inpsseq_mutation_ddg - Sheet2.csv\")\n",
    "demask_tuple_sub_map = {}\n",
    "for _, _row in demask_sub_matrix.iterrows():\n",
    "    demask_tuple_sub_map[(_row[\"pos\"]-1, _row[\"WT\"], _row[\"var\"])] = _row[\"score\"]\n",
    "    demask_tuple_sub_map[(_row[\"pos\"]-1, _row[\"var\"], _row[\"WT\"])] = _row[\"score\"]\n",
    "\n",
    "# 1.2 – Conduct matrix substitution\n",
    "#   --> First we create a tuple that has the wildtype amino acid and the \n",
    "#       mutant amino acid to access the substitution matrix\n",
    "#   --> Second we access the substitution matrix and replace with the respective score\n",
    "#       and in cases where no respective score is found we mark it to be updated later\n",
    "test_df[\"demask_sub_matrix_tuple\"] = test_df[[\"edit_idx\", \"wildtype_aa\", \"mutant_aa\"]].apply(tuple, axis=1)\n",
    "test_df[\"demask_sub_matrix_score\"] = test_df[\"demask_sub_matrix_tuple\"].progress_apply(lambda _mutant_tuple: demask_tuple_sub_map.get(_mutant_tuple, \"tbd\"))\n",
    "\n",
    "# 1.3 – Fill in the missing data with default values for now\n",
    "#   --> We aim to make this as similar to the -10.0 that was previously used as possible\n",
    "#       the most similar value considering the different scale would be --> -0.75\n",
    "#   --> No change remains as 0.0\n",
    "DEMASK_DEFAULT__DELETION__MATRIXSCORE = -0.75 # This is about 1.25 times larger than the minimum value (just like -10 in the original)\n",
    "DEMASK_DEFAULT__NO_CHANGE__MATRIXSCORE = 0.0\n",
    "test_df.loc[test_df['edit_type']==\"deletion\", 'demask_sub_matrix_score'] = DEMASK_DEFAULT__DELETION__MATRIXSCORE\n",
    "test_df.loc[test_df['edit_type']==\"no_change\", 'demask_sub_matrix_score'] = DEMASK_DEFAULT__NO_CHANGE__MATRIXSCORE\n",
    "test_df[\"demask_sub_matrix_score\"] = test_df[\"demask_sub_matrix_score\"].astype(float)\n",
    "\n",
    "# 1.4 – Calculate the column rank similar to what we did previously\n",
    "test_df['demask_sub_matrix_score_rank'] = stats.rankdata(test_df['demask_sub_matrix_score'])\n",
    "\n",
    "# 1.5 – Plot the delta between the assigned rank values between the two approaches for raw sub_matrix_score\n",
    "plt.figure(figsize=(18, 6))\n",
    "(test_df[\"sub_matrix_score_rank\"]-test_df[\"demask_sub_matrix_score_rank\"]).abs().plot(kind=\"hist\", bins=100)\n",
    "plt.title(\"Disagreement Amount Between DeMaSk and BloSUM100\", fontweight=\"bold\")\n",
    "plt.xlabel(\"Delta Change in Rank Between DeMaSk & BloSUM100\", fontweight=\"bold\")\n",
    "plt.ylabel(\"Frequency of Occurence\", fontweight=\"bold\")\n",
    "plt.grid(which=\"both\")\n",
    "plt.show()\n",
    "\n",
    "# 1.6 – Identify the columns to be combined\n",
    "combo_cols = ['b_factor_rank', 'demask_sub_matrix_score_rank', 'ddg_rank']\n",
    "\n",
    "# 1.7 – Combine the columns by multiplying them together\n",
    "test_df[\"demask_combined_val\"] = test_df[combo_cols].apply(np.prod, axis=1)\n",
    "\n",
    "# 1.7/ – To be fully added later... convert both outputs to rank\n",
    "test_df[\"combined_val_rank\"] = stats.rankdata(test_df['combined_val'])\n",
    "test_df[\"deamsk_combined_val_rank\"] = stats.rankdata(test_df['demask_combined_val'])\n",
    "\n",
    "# 1.8 – Plot the newly created column to see the distribution of values\n",
    "#       prior to any type of manipulation (i.e. raising to 1/3 power)\n",
    "plt.figure(figsize=(14,6))\n",
    "test_df[\"demask_combined_val\"].hist()\n",
    "plt.title(\"Distribution of Values Prior To Being Raised to 1/3 Power\")\n",
    "plt.xlabel(\"b_factor_rank * demask_sub_matrix_score_rank * ddg_rank\", fontweight=\"bold\")\n",
    "plt.ylabel(\"frequency of occurence in dataset\", fontweight=\"bold\")\n",
    "plt.show()\n",
    "\n",
    "# 1.9 – 'Normalize' the combined value column by raising to a particular power (as we are using Spearman this does nothing)\n",
    "COMBO_NORM_PWR = 1/3\n",
    "test_df[\"demask_norm_combined_val\"] = test_df[\"demask_combined_val\"]**COMBO_NORM_PWR\n",
    "\n",
    "# 1.10 – Plot the newly created column to see the distribution of values\n",
    "#       after manipulation (i.e. raising to 1/3 power)\n",
    "plt.figure(figsize=(14,6))\n",
    "test_df[\"demask_norm_combined_val\"].hist()\n",
    "plt.title(\"Distribution of Combined Normalized Values After Being Raised to 1/3 Power\")\n",
    "plt.xlabel(\"(b_factor_rank * demask_sub_matrix_score_rank * ddg_rank)^(1/3)\", fontweight=\"bold\")\n",
    "plt.ylabel(\"frequency of occurence in dataset\", fontweight=\"bold\")\n",
    "plt.show()\n",
    "\n",
    "# 1.11 – Plot the delta between the assigned rank values between the two approaches after combination and normalization\n",
    "plt.figure(figsize=(18, 6))\n",
    "(test_df[\"norm_combined_val\"]-test_df[\"demask_norm_combined_val\"]).abs().plot(kind=\"hist\", bins=100)\n",
    "plt.title(\"Disagreement Amount Between DeMaSk and BloSUM100\", fontweight=\"bold\")\n",
    "plt.xlabel(\"Delta Change in Rank Between DeMaSk & BloSUM100\", fontweight=\"bold\")\n",
    "plt.ylabel(\"Frequency of Occurence\", fontweight=\"bold\")\n",
    "plt.grid(which=\"both\")\n",
    "plt.show()\n",
    "\n",
    "# 1.12 – Create the submission file\n",
    "demask_seqid_2_tmrank_mapping = test_df.groupby(\"seq_id\")[\"demask_norm_combined_val\"].first().to_dict()\n",
    "ss_df[\"tm\"] = ss_df[\"seq_id\"].apply(lambda x: demask_seqid_2_tmrank_mapping[x])\n",
    "ss_df.to_csv(\"demask_submission.csv\", index=False)\n",
    "\n",
    "# 1.13 – Display the sample submission dataframe and it's details\n",
    "display(ss_df.describe().T)\n",
    "display(ss_df)\n",
    "\n",
    "# 1.14 – Display the updated test dataframe (and describe float/int based columns)\n",
    "display(test_df.describe().T)\n",
    "display(test_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1.11 – Plot the delta between the assigned rank values between the two approaches after combination and normalization\n",
    "plt.figure(figsize=(18, 6))\n",
    "(test_df[\"demask_combined_val\"]-test_df[\"demask_norm_combined_val\"]).abs().plot(kind=\"hist\", bins=100)\n",
    "plt.title(\"Disagreement Amount Between DeMaSk and BloSUM100\", fontweight=\"bold\")\n",
    "plt.xlabel(\"Delta Change in Rank Between DeMaSk & BloSUM100\", fontweight=\"bold\")\n",
    "plt.ylabel(\"Frequency of Occurence\", fontweight=\"bold\")\n",
    "plt.grid(which=\"both\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cols_to_ensemble = [\"ddg_rank\", \"b_factor_rank\", \"sub_matrix_score_rank\", \"demask_sub_matrix_score_rank\", ]\n",
    "ss_df[\"tm\"] = test_df[cols_to_ensemble].mean(axis=1)\n",
    "ss_df.to_csv(\"submission.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_df[test_df[\"seq_id\"]==32559][[_c for _c in test_df.columns if 'rank' in _c]].T.rename(columns={1169:\"wildtype_relative_ranking\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 5.1 – Define a function to return the substitution matrix (backwards and forwards)\n",
    "from Bio.SubsMat import MatrixInfo\n",
    "def get_sub_matrix(matrix_name=\"blosum100\"):\n",
    "    sub_matrix = getattr(MatrixInfo, matrix_name)\n",
    "    sub_matrix.update({(k[1], k[0]):v for k,v in sub_matrix.items() if (k[1], k[0]) not in list(sub_matrix.keys())})\n",
    "    return sub_matrix\n",
    "sub_matrix = get_sub_matrix()\n",
    "\n",
    "# 5.2 – Conduct matrix substitution\n",
    "#   --> First we create a tuple that has the wildtype amino acid and the \n",
    "#       mutant amino acid to access the substitution matrix\n",
    "#   --> Second we access the substitution matrix and replace with the respective score\n",
    "#       and in cases where no respective score is found we mark it to be updated later\n",
    "test_df[\"sub_matrix_tuple_asym\"] = test_df[[\"mutant_aa\", \"wildtype_aa\"]].apply(tuple, axis=1)\n",
    "test_df[\"sub_matrix_score_asym\"] = test_df[\"sub_matrix_tuple_asym\"].progress_apply(lambda _mutant_tuple: sub_matrix.get(_mutant_tuple, \"tbd\"))\n",
    "\n",
    "# 5.3 – Fill in the missing data with default values for now\n",
    "#   --> We set the default value for matrix sub to be equivalent to the bottom quartile value\n",
    "#       of all substitutions... this is because it is more deleterious than simple substitutions (larger difference)\n",
    "#   --> The default no_change value is 1 higher than the max score because a higher score means more similarity\n",
    "#DEFAULT__DELETION__MATRIXSCORE  = test_df[test_df[\"edit_type\"]==\"substitution\"][\"sub_matrix_score\"].quantile(q=0.25)\n",
    "DEFAULT__DELETION__MATRIXSCORE = -10.0\n",
    "#DEFAULT__NO_CHANGE__MATRIXSCORE = test_df[test_df[\"edit_type\"]==\"substitution\"][\"sub_matrix_score\"].max()+1.0\n",
    "DEFAULT__NO_CHANGE__MATRIXSCORE = 0.0\n",
    "test_df.loc[test_df['edit_type']==\"deletion\", 'sub_matrix_score_asym'] = DEFAULT__DELETION__MATRIXSCORE\n",
    "test_df.loc[test_df['edit_type']==\"no_change\", 'sub_matrix_score_asym'] = DEFAULT__NO_CHANGE__MATRIXSCORE\n",
    "test_df[\"sub_matrix_score_asym\"] = test_df[\"sub_matrix_score_asym\"].astype(float)\n",
    "test_df[\"sub_matrix_score_asym_rank\"] = stats.rankdata(test_df[\"sub_matrix_score_asym\"])\n",
    "\n",
    "# 5.4 – Display the updated dataframe (and describe float/int based columns)\n",
    "display(test_df.describe().T)\n",
    "display(test_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cols_to_ensemble = [\"ddg_rank\", \"ddg_rank\", \"ddg_rank\", \"b_factor_rank\", \"b_factor_rank\", \"b_factor_rank\", \"sub_matrix_score_rank\", \"sub_matrix_score_rank\", \"sub_matrix_score_asym_rank\", \"demask_sub_matrix_score_rank\", \"demask_sub_matrix_score_rank\", \"demask_sub_matrix_score_rank\"]\n",
    "ss_df[\"tm\"] = test_df[cols_to_ensemble].mean(axis=1)\n",
    "ss_df.to_csv(\"submission.csv\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.13 ('ai')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "e8e21a823b9237d214abb9e45b35d2642c53c6fa0410d12d295e0b957bff964d"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
